chain/chain/src/store.rs

```diff
@@ -142,6 +142,20 @@ pub trait ChainStoreAccess {
     fn get_block_header(&mut self, h: &CryptoHash) -> Result<&BlockHeader, Error>;
     /// Returns hash of the block on the main chain for given height.
     fn get_block_hash_by_height(&mut self, height: BlockHeight) -> Result<CryptoHash, Error>;
+    /// Returns hash of the first available block after genesis.
+    fn get_earliest_block_hash(&mut self) -> Result<Option<CryptoHash>, Error> {
+        let head_header_height = self.head_header()?.height();
+        let tail = self.tail()?;
+        if let Ok(block_hash) = self.get_block_hash_by_height(tail) {
+            return Ok(Some(block_hash.clone()));
+        }
+        for height in tail + 1..=head_header_height {
+            if let Ok(block_hash) = self.get_block_hash_by_height(height) {
+                return Ok(Some(self.get_block_header(&block_hash)?.prev_hash().clone()));
+            }
+        }
+        Ok(None)
+    }
```

Comment:
This is the core of the PR (the rest is just renaming of BlockIdOrFinality to BlockReference as the original naming does not fit the extended semantic).